### ISSUE TITLE
[feature request] [dataloader] Introduce Dataset.__collate__

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#33181]
And use it if collate_fn is unspecified and dataset.__collate__ is specified. This makes sense, because collation logic is very often tied to __getitem__ return values interface

**Comment by SsnL** [comment]
:) Maybe also `__worker_init__`?

**Comment by vadimkantorov** [comment]
@SsnL I understand, this could go far :) On serious note, I think that consuming these methods from the dataset simplifies distribution of the dataset class (in torchvision or torchaudio etc): i.e. the users don't need to remember to put in the good collate function and can just fire up the data loader. Especially for functions tightly-coupled with getitem impl. But of course this is more of a UX/api design issue, not very critical.

**Comment by ezyang** [comment]
I'm not convinced it should be an underscore-underscore method, but this certainly seems reasonable enough.

**Comment by gchanan** [comment]
@SsnL what's your take?

**Comment by SsnL** [comment]
I think this is a nice feature to have. The double underscore sounds fine to me, as it would have less BC issue and makes sense as a python API. 

However, I do wonder if we should use another word than `collate`. Now that the data loader can yield individual non-batched samples, I wonder if we can have a reasonable API name that captures both cases.

For `__worker_init__`, I have similar concerns. When we implement worker-reusing, There necessarily has to be an `iter_init` or `epoch_init`. I wonder if what are good API names for these init functions.

**Comment by SsnL** [comment]
How does an interface like this sound?
```py
class DatasetABC(abc.ABC):
    def __worker_init__(self):
        pass
        
    def __epoch_init__(self):
        pass
        
    def __collate__(self, batch):
        # For BC, DataLoader will still test 
        #   if hasattr(dataset, '__collate__'):
        #       collate_fn = dataset.__collate__
        #   else:
        #       collate_fn = default_collate
        return default_collate(batch)
        
    def __convert__(self, element):  # for non-batching loading
        # For BC, DataLoader will still test 
        #   if hasattr(dataset, '__convert__'):
        #       convert_fn = dataset.__convert__
        #   else:
        #       convert_fn = default_collate
        return default_convert(element)

    def __len__(self):
        return NotImplemented


class Dataset(DatasetABC):
    r"""Map-style"""
    @abc.abstractmethod
    def __getitem__(self, key):
        return NotImplemented


class IterableDataset(DatasetABC):
    r"""iterable-style"""
    @abc.abstractmethod
    def __iter__(self):
        return NotImplemented
```

**Comment by vadimkantorov** [comment]
LGTM :) Though, maybe @ezyang is right about underscores, since it becomes less "service" methods, but more "logic" methods.

Also, how about `worker_teardown`/`worker_terminate`?

**Comment by SsnL** [comment]
I agree. Although the problem with no underscores is that it may have backward compatibility issues...

**Comment by vadimkantorov** [comment]
One detail: your API of __default_collate__ and __convert__ accepts self. There won't be any double-copies of dataset object because of this, right? These are executed on the main thread, not on worker threads?

For __worker_init__, will this self arg incur a double copy/pickle?

**Comment by SsnL** [comment]
`collate` and `convert` are executed on the worker. 

There won't be double copies because these are just regular python methods on python class. Python pickling hands it well.

**Comment by vadimkantorov** [comment]
For double copy, I was afraid that the dataset object will be pickled separately from dataset.collate. If this the case, there should be double pickling, right? or it would figure out that reference is the same?

In any case this should be easy enough to check

**Comment by SsnL** [comment]
Python pickling xreferences. So if you send `[obj_A, obj_A]`, you won't get `[obj_A_copy1, obj_A_copy2]`. Similarly for classes with methods.

**Comment by vadimkantorov** [comment]
Okay, got it. So what you are saying is that they would be pickled together, not separately. Then everything should be okay :)

**Comment by SsnL** [comment]
Yes, exactly!

**Comment by Haydnspass** [comment]
@SsnL Thanks for pointing me to this! :)

The issue with your proposal above is however that it won't work recursively because default_collate is a function. 

So suppose you want to return a tuple (FancyTypeA, FancyTypeB) and implement the collate such that when the element is FancyTypeA/B it will aggregate the elements to a list.
```
...
 if isinstance(elem, [FancyTypeA, FancyTypeB]):
    return [e for e in batch]
```

But now when __collate__ sees a list/tuple it will default to the default_collate function and from within the default collate it can not escape to the implementation of the collate of FancyTypeA/B. 
The workaround would be to implement a collate for lists/tuples as well but this would be redundant because it's already there in the default.

**Comment by SsnL** [comment]
@Haydnspass we can make this recursive by :)

```py
class Dataset:
    def __collate__(self, samples):
       # content of default_collate but recursively calls `self.__collate__`
```

Then subclass can easily 
    def __collate__(self, samples):
        if isinstance(samples[0], MyType):
             # do my thing
        else:
            return super().__collate__(samples)
       # content of default_collate but recursively calls `self.__collate__`

**Comment by Haydnspass** [comment]
Indeed! I agree it's the better solution :)

**Comment by SsnL** [comment]
In #48536 , @freud14 @vadimkantorov and I are talking about actually implementing these. It would be great to hear thoughts from @ezyang @gchanan and/or @VitalyFedyunin !

**Comment by VitalyFedyunin** [comment]
Replied inside #48536

**Comment by Haydnspass** [comment]
Is there any update on this? Any strategies on how to move forward?
It is a bit unfortunate that the current default is still functionally designed and therefore hard to extend.
